File: breadth_first_traversal/BreadthFirstTraversal.java
import java.util.HashSet;
import java.util.Iterator;
import java.util.LinkedList;
import java.util.Set;

public class BreadthFirstSearch {

    // Array  of lists for Adjacency List Representation
    public static LinkedList<Integer>[] adj;

    //Function to add an edge into the breadthFirstTraversal
    private static void addEdge(int source, int w) {
        adj[source].add(w);
    }

    private static void breadthFirstTraversal(int source, int destination) {
        // false by default in java)
        Set<Integer> visited = new HashSet();
        // Mark the current node as visited
        visited.add(source);
        LinkedList<Integer> queue = new LinkedList<>();
        queue.add(source);
        while (queue.size() != 0) {
            source = queue.poll();
            System.out.println(source);
            Iterator<Integer> i = adj[source].listIterator();
            int flag = 0;
            while (i.hasNext()) {
                int n = i.next();
                if (n == destination) {
                    System.out.println(n);
                    flag = 1;
                    break;
                }
                if (!visited.contains(n)) {
                    visited.add(n);
                    queue.add(n);
                }
            }
            if (flag == 1) {
                break;
            }
        }
    }   

    public static void initEdges(int n) {
        adj = new LinkedList[n];
        for (int i = 0; i < n; ++i) {
            adj[i] = new LinkedList();
        }
    }

    public static void main(String[] args) {
        initEdges(4);
        addEdge(0, 1);
        addEdge(0, 3);
        addEdge(1, 2);
        addEdge(2, 1);
        addEdge(2, 3);
        addEdge(3, 3);
        System.out.println("Breadth First Traversal starting from source to destination");
        breadthFirstTraversal(1,3);
    }
}

File: breadth_first_traversal/breadth_first_traversal.py
"""
Breadth-first-traversal is an algorithm for traversing a tree or
graph data structure. Starting at the tree root (or some arbitrary node of a
graph, sometimes referred to as a 'search key'[1]) and explores the neighbor
nodes at that level first, before moving to the next level.
"""

from collections import deque


def breadth_first_traversal(graph, source):
    """ Performs a breadth-first traversal on a graph

    Args:
        graph (list of list of int): Adjacency matrix representation of graph
        source (int): Index of source vertex to begin search from

    Returns:
        list of dicts describing each vertex in the searched graph
            -> [{distance: _, predecessor: _ }]
    """
    vertex_info = []
    for i in range(len(graph)):
        vertex_info.append({"distance": None, "predecessor": None})
    vertex_info[source]["distance"] = 0

    search_queue = deque()
    search_queue.append(source)

    while search_queue:
        u = search_queue.popleft()
        for v in graph[u]:
            if vertex_info[v]["distance"] is None:
                vertex_info[v]["distance"] = vertex_info[u]["distance"] + 1
                vertex_info[v]["predecessor"] = u
                search_queue.append(v)
    return vertex_info


def main():
    graph_adj_list = [
        [1],
        [0, 4, 5],
        [3, 4, 5],
        [2, 6],
        [1, 2],
        [1, 2, 6],
        [3, 5],
        []
    ]
    vertex_info = breadth_first_traversal(graph_adj_list, 3)

    for i in range(len(graph_adj_list)):
        print("vertex %s : distance = %s, predecessor = %s" %
              (i, vertex_info[i]["distance"], vertex_info[i]["predecessor"]))

    assert(vertex_info[0] == {
        "distance": 4,
        "predecessor": 1
    })
    assert(vertex_info[1] == {
        "distance": 3,
        "predecessor": 4
    })
    assert(vertex_info[2] == {
        "distance": 1,
        "predecessor": 3
    })
    assert(vertex_info[3] == {
        "distance": 0,
        "predecessor": None
    })
    assert(vertex_info[4] == {
        "distance": 2,
        "predecessor": 2
    })
    assert(vertex_info[5] == {
        "distance": 2,
        "predecessor": 2
    })
    assert(vertex_info[6] == {
        "distance": 1,
        "predecessor": 3
    })
    assert(vertex_info[7] == {
        "distance": None,
        "predecessor": None
    })


if __name__ == '__main__':
    main()
